ambry-commons/src/main/java/com/github/ambry/commons/PipedAsyncWritableChannel.java

```diff
@@ -334,11 +348,12 @@ private Callback<Long> makeCallbackForReadableStreamChannel(boolean primary, Ato
               }
               resolveChunk(primaryWriteCallbackResult.bytesWritten, primaryWriteCallbackResult.exception);
             } else if (primaryWriteCallbackResult != null) {
-              // TODO: Write successful callback came from primary but not from secondary. We will "start a timer" to wait
-              //  for result from secondary. If we time out waiting for the result, we will close secondary and send the
-              //  remaining bytes to primary alone so that primary upload SLA is not affected.
-            } else {
-              // Do nothing. Callback from Secondary came. We will wait for callback to come from primary
+              secondaryTimeout = wheel.newTimeout(timeout -> {
```

Comment:
HashedWheelTimer actually limit the number of concurrent instances to 64, so we create one instance per Pipe. We need to share the instance between different pipes. It would be ideal to only create one instance for all pipes, but we don't really know if one instance would be good enough.

I suggest that we have a metric to keep track of the delayed timeout duration and see if one instance would have enough capacity to handle all timeouts. 

For instance, when we summit a timeout task to wheel, we also would know the expected timeout time. And when the timeout callback is finally invoked, we can compare it with the expected timeout value and see if the difference. Let's have a histogram of the different in ms. If one instance can deal with all the timeout, then we should see the 95% of difference to be always be within one tick. If not, it means we need to create more instances and randomly assign one instance to one pipe.

Reply:
Sure. Added metric to keep track of delayed timeout duration